### ISSUE TITLE
PackageInternalizer - Cannot Process String in URL

### ISSUE THREAD

**Issue by ryanrichter94** [chocolatey/chocolatey-licensed-issues#177]
# Issue 

Cannot internalize a package that has a string in the download URL path

### Example

$url = "https://s3.amazonaws.com/ddagent-windows-stable/ddagent-cli-$($env:chocolateyPackageVersion).msi"

### Choco Error Response 

Unable to find value for variable '$($env'. May not be able to download resource from url 'https://s3.amazonaws.com/ddagent-windows-stable/ddagent-cli-$($env:chocolateyPackageVersion).msi'.

Download doesn't yet support replacement of string methods. May not be able to download resource from url 'https://s3.amazonaws.com/ddagent-windows-stable/ddagent-cli-$($env:chocolateyPackageVersion).msi'.

### References

* [Internal Issue](https://gitlab.com/chocolatey/collaborators/choco-licensed/-/issues/354)
* [ZenDesk Ticket](https://chocolatey.zendesk.com/agent/tickets/7945)

**Comment by ferventcoder** [comment]
How it works when it is not an environment variable:

![image](https://user-images.githubusercontent.com/63502/92759438-5a3df100-f355-11ea-864f-1b0c3fad26dc.png)

**Comment by vexx32** [comment]
Yeah, looks like we're probably regexing things... and we should not be regexing variables, period. We should lean as much as possible on our PowerShell hooks to find where variables are and replace them if we need to do that manually.

**Comment by pauby** [comment]
@shawnfunke The issue has not been assigned to a milestone, so it's not something we are working on just now.

**Comment by pauby** [comment]
@shawnfunke We do have another release of Chocolatey Licensed Extension _planned_ this year (after the upcoming 6.0.0 release). But even with the best will in the world, plans do change.

I'm not sure that helps and I apologise for that.

**Comment by Stunkymonkey** [comment]
@shawnfunke a current workaround for me: `choco download prometheus-windows-exporter.install --no-progress --version 0.22.0 --internalize-all-urls --internalize version=0.22.0`

**Comment by shawnfunke** [comment]
@Stunkymonkey what you mentioned isn't a workaround, there's been a fix regarding the package version substitution in released version `6.1.0` [^1]. After a quick test I can confirm that this has solved the above mentioned issue.

@pauby I'm not sure if there is anything else regarding this issue, if not I'd be fine with marking this issue as resolved. 

[^1]: https://docs.chocolatey.org/en-us/licensed-extension/release-notes#june-29-2023

**Comment by pauby** [comment]
@shawnfunke I don't think this has been fixed as it's not in the 6.1.0 milestone and therefore completed. Just testing it on a dummy package, with Chocolatey Licensed Extension 6.1.0, I get:

```
Chocolatey v2.1.0 Business
Downloading existing package(s) to C:\tmp\snagit\snagit\internal\download

snagit v2022.1.2.20221010 (forced)
Found internalizable Chocolatey functions. Inspecting values for remote resources.
Unable to find value for variable '$($env'. May not be able to download resource from url 'https://download.techsmith.com/snagit/releases/$($env:ChocolateyPackageVersion)/snagit.msi'.
Download doesn't yet support replacement of string methods. May not be able to download resource from url 'https://download.techsmith.com/snagit/releases/$($env:ChocolateyPackageVersion)/snagit.msi'.
Found 'https://download.techsmith.com/snagit/releases/$($env:ChocolateyPackageVersion)/snagit.msi' for replacement.
```

**Comment by OraDotNetDev** [comment]
I was wondering whether the Internalizer's processing of the [ChromeDriver community package ](https://community.chocolatey.org/packages/chromedriver/116.0.5845.962) is also covered by this issue?

When the Internalizer executes against [ChromeDriver](https://github.com/yetanotherchris/Choco-chromedriver/issues/12), the Json manifest (listing installation payloads) referenced within chocolateyinstall.ps1 is captured and added to the recompiled package, however not the target payload/archive ie `https://edgedl.me.gvt1.com/edgedl/chrome/chrome-for-testing/116.0.5845.96/win32/chrome-win32.zip` (see output below.)

Thanks

OraDotNetDev

```
# chromedriver
# chocolateyinstall.ps1

$jsonUrl = "https://googlechromelabs.github.io/chrome-for-testing/known-good-versions-with-downloads.json"
# Fetch the JSON content from the URL
$jsonContent = Invoke-RestMethod -Uri $jsonUrl
# Find the desired versioninfo
$desiredVersionInfo = $jsonContent.versions | Where-Object { $_.version -eq $chromedriverVersion }

# Find the Chromedriver download URL for the desired platform
$chromedriverUrl = $desiredVersionInfo.downloads.$packageName | 
    Where-Object { $_.platform -eq $desiredPlatform } | 
    Select-Object -ExpandProperty url

Install-ChocolateyZipPackage "packageName" -url "$chromedriverUrl" -unzipLocation "$toolsDir" -checksumType "sha256" -checksum "$checkSum"
```

```
PS C:\> choco download chromedriver --source='C:\Temp\Source' --internalize -i
Chocolatey v2.2.2 Business
Downloading existing package(s) to C:\download

chromedriver v116.0.5845.96
Found internalizable Chocolatey functions. Inspecting values for remote resources.
Found 'https://googlechromelabs.github.io/chrome-for-testing/known-good-versions-with-downloads.json' for replacement.
Downloading: 100% - 349.28 KB
Download of 'known-good-versions-with-downloads.json' completed.
Replacing 'https://googlechromelabs.github.io/chrome-for-testing/known-good-versions-with-downloads.json' with '$(Split-Path -parent $MyInvocation.MyCommand.Definition)\files\known-good-versions-with-downloads.json'

Adding -UseOriginalLocation to Install-ChocolateyPackage (if found).
 You can also turn off the feature 'internalizeAppendUseOriginalLocation' to not append
Updating chocolateyInstall.ps1 with local resources.
Recompiling package...
```

**Comment by pauby** [comment]
@OraDotNetDev While fixing this issue may fix the internalization of that package, we don't support any packages on the Chocolatey Community Repository. If the package can be amended to work with the current internalization, it would be best to get it amended by contacting the maintainer directly.
